src/Utility/Hash.php

```diff
@@ -1146,6 +1146,86 @@ public static function mergeDiff(array $data, array $compare): array
         return $data;
     }
 
+    /**
+     * Validates data types for each key in $data against $types.
+     *
+     * Every key in $data must have a corresponding key in $types
+     * and the value in $data must match the expected type.
+     *
+     * Keys with mismatched or missing types are returned.
+     *
+     * Types can be one of the following or an array of possible types:
+     *
+     *  - 'null'
+     *  - 'array'
+     *  - 'string'
+     *  - 'float'
+     *  - 'int'
+     *  - 'bool'
+     *
+     *  If not one of the above built-in types, performs an `instanceof` check against the type.
+     *
+     * @param array $data The data to validate
+     * @param array $types The types for each expected key
+     * @param bool $strict Require all keys in both $data and $types to match
+     * @return array
+     */
+    public static function validate(array $data, array $types, bool $strict = false): array
```

Comment:
I like the general idea of being able validate the 'shape' of an array. Should we consider using a format like jsonschema or is that too heavy handed? Right now a limitation I can see is that this can't be used for nested arrays which we have a bit of in our configuration data.

Reply:
Nested arrays would require nested types. At the moment, I am using arrays to store more than one allowed value: `['string', 'bool'].  That would need to change into something more structured. It's something I want to support, but stopped before getting feedback.